loopy/transform/dependency.py

```diff
@@ -0,0 +1,240 @@
+        amf(insn.expression, insn.id)
+
+    # determine transitive dependencies between statements
+    transitive_deps = {}
+
+    deps_dag = {insn.id: insn.depends_on for insn in knl.instructions}
+
+    from pytools.graph import compute_topological_order
+    t_sort = compute_topological_order(deps_dag)
+
+    for insn_id in t_sort:
+        transitive_deps[insn_id] = reduce(
+                frozenset.union,
+                (transitive_deps.get(dep, frozenset([dep]))
+                      for dep in
+                      knl.id_to_insn[insn_id].depends_on),
+                frozenset())
+
+    # compute and store precise dependencies
+    new_insns = []
+    for insn in knl.instructions:
+        happens_after = insn.happens_after
+
+        # get access maps
+        for dependency, variable in happens_after:
+            assert isinstance(insn.id, str) # stop complaints
+
+            after_map = amf.get_map(insn.id, variable)
+            before_map = amf.get_map(dependency, variable)
+
+            # scalar case(s)
+            if before_map is None or after_map is None:
+                warn("unable to determine the access map for %s. Defaulting to "
+                     "a conservative dependency relation between %s and %s" %
+                     (dependency, insn.id, dependency))
+
+                # clean up any deps that do not contain a variable name
+                happens_after = {insn_id: dep
+                                 for insn_id, dep in happens_after.items()
+                                 if dep.variable_name is not None}
+
+                continue
+
+            dims = [before_map.dim(dim_type.out),
+                    before_map.dim(dim_type.in_),
+                    after_map.dim(dim_type.out),
+                    after_map.dim(dim_type.in_)]
+
+            for i in range(len(dims)):
+                if i == 0 or i == 1:
+                    scalar_insn = dependency
+                else:
+                    scalar_insn = insn.id
+
+                if dims[i] == 0:
+                    warn("found evidence of a scalar access in %s. Defaulting "
+                         "to a conservative dependency relation between "
+                         "%s and %s" % (scalar_insn, dependency, insn.id))
+
+                    # clean up any deps that do not contain a variable name
+                    happens_after = {insn_id: dep
+                                     for insn_id, dep in happens_after.items()
+                                     if dep.variable_name is not None}
+
+                    continue
+
+            # non-scalar accesses
+            dep_map = after_map.apply_range(before_map.reverse())
+            dep_map -= dep_map.identity(dep_map.get_space())
+
+            if insn.happens_after[dependency] is not None:
+                lex_map = insn.happens_after[dependency].instances_rel
+            else:
+                lex_map = dep_map.lex_lt_map(dep_map)
```

Comment:
We should use the fine-grain/instance-level dependency info that's already in the kernel at this point.